bottlecap/src/telemetry/listener.rs

```diff
@@ -167,86 +203,91 @@ impl TelemetryListener {
     /// # Errors
     ///
     /// Function will error if the passed address cannot be bound.
-    pub fn run(
+    pub async fn new(
         config: &TelemetryListenerConfig,
         event_bus: Sender<Vec<TelemetryEvent>>,
-    ) -> Result<TelemetryListener, Box<dyn Error>> {
+        cancel_token: tokio_util::sync::CancellationToken,
+    ) -> Result<TelemetryListener, String> {
         let addr = format!("{}:{}", &config.host, &config.port);
-        let listener = TcpListener::bind(addr)?;
-
-        let join_handle = std::thread::spawn(move || {
-            debug!("Initializing Telemetry Listener");
-
-            loop {
-                for stream in listener.incoming() {
-                    debug!("Received a Telemetry API connection");
-
-                    let cloned_event_bus = event_bus.clone();
-                    if let Ok(stream) = stream {
-                        std::thread::spawn(move || {
-                            let r = Self::handle_stream(&stream, cloned_event_bus);
-                            if let Err(e) = Self::acknowledge_request(stream, r) {
-                                error!("Error acknowledging Telemetry request: {:?}", e);
-                            }
-                        });
-                    } else {
-                        error!("Error accepting connection");
+        let listener = TcpListener::bind(addr).await.map_err(|e| e.to_string())?;
+
+        Ok(TelemetryListener {
+            event_bus,
+            cancel_token,
+            listener,
+        })
+    }
+
+    pub async fn spin(self) {
+        loop {
+            tokio::select! {
+                biased;
+                stream = self.listener.accept() => {
+                    match stream {
```

Comment:
Is it possible to remove this assignment by just doing it inline?

Reply:
not that I'm aware of? Tokio wants a future to block on and this is how you assign the result of that future. Possibly we could refactor this all into another method but I think we'll pass on that